[WIP][JIT] Add JIT backend registration API

SplitInfinity opened the pull request:
**Summary**
This commit adds `torch::jit::RegisterBackend`, an API that allows
external backends to be registered for the execution of JIT subgraphs
outside the JIT interpreter. In order to register an external backend,
one must extend the provided abstract class `PyTorchBackendInterface` and provide
two additional functions: one that creates an instance of the aforementioned subclass
of `PyTorchBackendInterface`, and another that preprocesses a `ScriptModule` so that
it can run on the backend. Then, a `ScriptModule` that can compile and execute a given
JIT subgraph using the functions provided at registration time is generated
for each registered backend.

**Testing**
This commit adds a unit test that uses a minimal test backend
to make sure that the registration endpoint and generated
`ScriptModule` work.

```
$ python test/test_jit.py TestBackends
Fail to import hypothesis in common_utils, tests are not derandomized
.
----------------------------------------------------------------------
Ran 1 test in 0.183s

OK

```

eellison committed:
[WIP][JIT] Add JIT backend registration API

Summary:
This commit adds `torch::jit::backend`, an API that allows
external backends to be registered for the execution of JIT subgraphs
outside the JIT interpreter. In order to register an external backend,
one must extend the provided abstract class `PyTorchBackendInterface` and provide
two additional functions: one that creates an instance of the aforementioned subclass
of `PyTorchBackendInterface`, and another that preprocesses a `ScriptModule` so that
it can run on the backend. Then, a `ScriptModule` that can compile and execute a given
JIT subgraph using the functions provided at registration time is generated
for each registered backend.

Testing:
This commit adds a unit test that uses a minimal test backend
to make sure that the registration endpoint and generated
`ScriptModule` work.

SplitInfinity commented:
I have marked this pull request as WIP (work-in-progress) because there are a few items I was unable to resolve myself and would like to discuss with others. I have made comments about these items inline.

SplitInfinity commented:
Also, this needs to be rebased, but I didn't do that yet because it's WIP anyway.

SplitInfinity commented on aten/src/ATen/core/type.cpp:
```diff
@@ -916,7 +916,7 @@ size_t ClassType::addAttribute(
   const char* what = is_parameter ? "parameter" : "attribute";
   checkNotExist(name, what);
-  checkNoAny(*this, what, name, type);
+  // checkNoAny(*this, what, name, type);
```
I disabled this check temporarily in order to register an `backend_handles` and `extra_infos` attributes of type `Dict[str, Any]` on the generated `ScriptModule` so as to allow backends to define what the backhand specific types look like. Is there a way to achieve this without declaring `Any` type attributes?

SplitInfinity replied:
Still unable to add attributes to the generated module that involve `Any` types without disabling this, but not sure if i can get rid of it altogether instead of commenting it out.

SplitInfinity replied:
I wonder if this is supposed to be a thing: https://circleci.com/gh/pytorch/pytorch/5302241

SplitInfinity commented on test/jit/test_backends.py:
```diff
@@ -0,0 +1,45 @@
+    def test_simple(self):
+        # Test compile.
+        scripted_module = torch.jit.script(MyModule())
+        lowered_module = torch.jit.to_test_backend(scripted_module._c, {"key": "value"})
```
FYI: I was not able to pass in an empty dict here due to a Tracer error.

SplitInfinity commented on test/jit/test_backends.py:
```diff
@@ -0,0 +1,45 @@
+        self.assertEqual(input, output)
+
+        # Test save and load.
+        # lowered_module.save("lowered.pt")
```
I think `__setstate__` and `__getstate__` are defined correctly, but the problem here is that attributes of type `Any` are not serializable.

driazati replied:
It's landed in #35741

SplitInfinity commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -29,7 +29,7 @@ struct Node;
 using ::c10::Symbol;
 
-using OperationCreator = Operation (*)(const Node*);
+using OperationCreator = std::function<Operation(const Node*)>;
```
I made this change in order to capture the `backend` struct in the lambdas registered as ops for `compile` and `execute`. I'd prefer not to make sure a wide-reaching change though, so please let me know if there are better alternatives.

SplitInfinity commented on torch/jit/__init__.py:
```diff
@@ -1342,6 +1342,10 @@ def wrapper(*args):
     return wrapper
 
 
+def to_test_backend(module, extra_info):
```
I'd prefer not to do this for every backend, so if anyone has a way to programmatically generate one of these methods for every backend, please let me know.

zdevito replied:
The intension in the design is that users do in fact write their own functions, to make it clear that bugs in the implementation should be submitted to the implementors of the backend and to as JIT bugs.

SplitInfinity commented on torch/csrc/jit/frontend/sugared_value.h:
```diff
@@ -284,8 +284,17 @@ struct TORCH_API BuiltinModule : public SugaredValue {
       // methods under its module.
       return std::make_shared<BuiltinModule>("aten", version);
     }
+
```
All of these changes are to make `torch.ops...` resolve correctly, which isn't possible with the `NativeResolver`. To be honest I kinda guessed, so I wouldn't be surprised if I missed something here.

SplitInfinity reviewed (commented):
Left some comments inline about things I want to discuss.

SplitInfinity commented:
cc: @yinghai @jamesr66a @jackm321 

Note that the PR is still in WIP state because there are some clowny things I would like to discuss with others.

SplitInfinity commented on test/cpp/jit/test_backends.cpp:
```diff
@@ -0,0 +1,46 @@
+// necessary to test that the JIT backend registration endpoints and
+// code generation are working correctly. It is not intended to
+// produce numerically correct results.
+RegisterBackend backend({
```
This backend is maybe *too* simple. I'd be open to modifying it so that it uses a `GraphExecutor` instance to run a graph.

SplitInfinity commented:
Also cc: @bzinodev too

yinghai commented:
High level question, can we just use torchbind for this?

yinghai commented on test/cpp/jit/test_backends.cpp:
```diff
@@ -0,0 +1,46 @@
+      push(stack, mod._ivalue());
+      return 0;
+    },
+    [/*compile=*/](Stack& stack) -> int {
```
Lift the comment outside?

yinghai commented on torch/csrc/jit/backends/backend.cpp:
```diff
@@ -0,0 +1,37 @@
+using BackendList = std::vector<Backend>;
+
+// Registry class for backends.
+struct BackendRegistry {
```
I'm reading https://github.com/pytorch/pytorch/pull/34140 and wonder if we can just use the custom class registry mechanism there? @jamesr66a

jamesr66a replied:
@yinghai yes, @suo and I have gone over the design and made changes to harmonize this with the custom classes API. Those adjustments will likely occur after this PR is landed

suo commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -69,7 +69,7 @@ struct TORCH_API Operator {
   Operator(
       const std::string& schema,
-      int (*op)(Stack&),
+      Operation op,
```
iirc, @zdevito was the one that made this constructor a non-capturing function pointer on purpose. @zdevito: can you take a look at this part of the file?

zdevito replied:
Yeah, this change is intentional. We don't want simple operators to store state because it makes it hard in the future to optimize the execution if we have to assume everything is more than a raw pointer.

SplitInfinity replied:
I need to undo this, it's no longer necessary.

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,126 @@
+          }
+
+          // Call compile to ensure that the returned Module is ready to run.
+          loweredModule.run_method("__compile__");
```
I think we should call `__setstate__` here directly, so that we test the exact path the `load()` will take.

zdevito commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,126 @@
+namespace torch {
+namespace jit {
+void initJitBackendBindings(PyObject* module) {
+  for (auto& backend : getAllBackends()) {
```
What if a backend gets registered after this is called? This will be the common case because backends will likely be in different modules from `torch`.

zdevito commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,126 @@
+    std::string execute_method_schema =
+        backend.name + "::execute(Any[] a) -> Any[]";
+
+    torch::jit::RegisterOperators reg({
```
These registrations won't need to exist when the object is bound in as a class rather than as builtins. I think that is the right solution to this problem. If we want to split this apart, I'd register these without schema for now to get around the Operation definition. 

I worry however, that landing this and then landing a future potential update to make it classes might cause a BC problem. We need to be clear that there is no BC for this version of to_backend until the change to classes happens.

zdevito commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,34 @@
+  std::string name;
+  /// A function that should be invoked to convert the JIT graph into
+  /// something that the backend can compile.
+  std::function<int(Stack&)> preprocess;
```
It would be clearer if these were virtual functions and users subclasses this.

zdevito commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,34 @@
+};
+
+// Static registration API for backends.
+struct TORCH_API RegisterBackend {
```
If these were bound in as classes then there would be no need to have a separate global registry of backends. Instead each backend could just register its class with the already existing class registry.

zdevito reviewed (commented):
I think most of the issues around registration goes away when we bind backends in as a custom class rather than operators. So the only thing to decide is whether to do that incrementally.

suo commented:
Yeah I think the plan was to land this PR as a starting point and for the next followup to be to switch over the registration mechanism to torchbind. The prospective users of this API know that it's not available yet to rely on in a bc-maintaining way.

SplitInfinity commented:
Hey everyone, I just pushed a new version after reviewing your comments and discussing this offline with @jamesr66a and @suo. The major changes are summarized below:
- The interface to a backend is now defined by an abstract class `PyTorchBackendInterface` that backends must extend.
- Backends must also define functions to preprocess a `ScriptModule` so that it can run on the backend and another function to create an appropriately configured backend instance (of the type that backends must define as per bullet #1 above). These functions are registered in the custom class method registry under the namespaces `torch.__backends__.__preprocess__` and `torch.__backends__.__creator__`.
- Consequently, there is no separate registry for registered backends, and the operators that enable the `ScriptModule` to call `compile` and `execute` are now stateless.

SplitInfinity commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -45,4 +45,25 @@ void registerCustomClassMethod(std::unique_ptr<jit::Function> fn) {
 }
 
+jit::Function* getCustomClassMethod(const c10::QualifiedName& name) {
+  for (auto& fn : customClassMethods()) {
```
Maybe time to use a `map` instead of a `vector`?

SplitInfinity commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -45,4 +45,25 @@ void registerCustomClassMethod(std::unique_ptr<jit::Function> fn) {
+
+std::vector<jit::Function*> getCustomClassMethodsWithPrefix(const c10::QualifiedName& prefix) {
+  std::vector<jit::Function*> result;
+  for (auto& fn : customClassMethods()) {
```
Prefix scan is linear time regardless of whether we use a `map` or `vector` but I don't think it's important to focus on optimizing this.

SplitInfinity commented on torch/jit/__init__.py:
```diff
@@ -1340,7 +1340,7 @@ def wrapper(*args):
 
     return wrapper
-
+    
```
Oops

SplitInfinity commented on torch/csrc/jit/backends/backend.cpp:
```diff
@@ -0,0 +1,58 @@
+      /*overload_name=*/"",
+      /*arguments=*/{mod},
+      /*returns=*/{capsule});
+  registerCustomClassMethod(std::make_unique<jit::BuiltinOpFunction>(
```
`registerCustomClassMethod` doesn't check for name collisions. We should probably do that.

jamesr66a commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,181 @@
+    std::string to_backend_method_name = "_jit_to_" + backend_name;
+    m.def(
+        to_backend_method_name.c_str(),
+        [=](Module orig_module, py::dict method_compile_spec) {
```
We might want to store this as a `std::function` somewhere so it can be accessed from C++ as well. That might also involve separating it from pybind11 (e.g. change `py::dict`). Don't know if that should be done on this PR, though.

cc @suo

jamesr66a commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,181 @@
+          loweredModule.define(
+              "\ndef __create_backend(self):\n\treturn ops.prim.backend_create(\"" +
+              backend_creator_name.qualifiedName() +
+              "\", self.__processed_module)\n");
```
To make this easier to read, can we use [CodeTemplate](https://github.com/pytorch/pytorch/blob/dbe850af5b941c6f0ae1ed50e7d6b779db55f5af/torch/csrc/jit/frontend/code_template.h#L97)?

SplitInfinity replied:
Thanks, didn't know about that, will look into it.

jamesr66a commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,181 @@
+            loweredModule.define(
+                "\ndef " + method_name +
+                "(self, input):\n\treturn ops.prim.backend_execute(self.__backend, self.__handles[\"" +
+                method_name + "\"], input)\n");
```
Same here, CodeTemplate if possible

SplitInfinity replied:
Thanks, didn't know about that, will look into it.

jamesr66a reviewed (commented):
Looking pretty good! I had some concerns about `execute()`and how in my reading it seems to only accept one argument. Could you let me know if something else is going on there?

SplitInfinity commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,51 @@
+// to create a type that represents a ScriptModule that has been preprocessed
+// so that it can run on the backend.
+class TORCH_API PyTorchBackendModuleInterface
+    : public torch::CustomClassHolder {};
```
It does, but something like this is necessary if the backend needs to expose its preprocessed module as a custom type. For example, let's say that Glow cannot store the preprocessed module as a `ScriptModule` for whatever reason, and needs to create its own type, `GlowModule` to store the information it needs. But if it is not a subclass of `torch::CustomClassHolder`, there is no way to construct an `IValue` to put it onto the stack (could be wrong here, but I don't see any other applicable constructor).

SplitInfinity commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,51 @@
+
+  // Execute the method corresponding to \p handle using \p input. \returns the
+  // output of the method.
+  virtual c10::IValue execute(c10::IValue handle, c10::IValue input) = 0;
```
I think it's easier for backend writers to write these functions when the function signatures are clear and well-defined, so I put the code that interfaces with the `Stack` into the primitives for `backend_compile` and `backend_execute`. But I guess we can't do varags then, or multiple inputs?

SplitInfinity commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,181 @@
+                PyTorchBackendInterface,
+                CustomClassHolder>(capsule);
+            TORCH_INTERNAL_ASSERT(backend);
+            auto res = backend->execute(handle, input);
```
I tried to add support for multiple arguments using `*args` in my codegenned `Module` but that didn't compile for some reason? Even if we turn `execute` into a function that takes a `Stack`, how do we pass the multiple args from Python to C++?

suo replied:
For vararg schemas, the interpreter will push the number of passed arguments onto the stack. So for `prim::Print("foo", "bar")`, the stack will look like `["foo", "bar", 2]`.

SplitInfinity commented:
Regarding backend functions as `Stack` based or not, I thought that having a concrete signature/schema would make things easier for backend writers, but if these `compile` and `execute` functions are `Stack` based, I think that allows us to do varargs for those and might actually be a prereq for multiple inputs once I find out why `*args` didn't work in my codegenned module.

I'll change that and re-push.

suo commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,181 @@
+            std::string method_name = py::cast<std::string>(e.first);
+
+            loweredModule.define(
+                "\ndef " + method_name +
```
We need some plumbing code to retrieve the schemas of each method from the original module and codegen the same one here instead. So the generated `forward` should look like:
```
def forward(self, foo, bar, keyword=argument):
     return ops.prim.backend_execute(self.__backend, self.__handles["forward"], foo, bar, keyword=keyword)
```

suo commented on test/jit/test_backends.py:
```diff
@@ -0,0 +1,50 @@
+
+        # Test execute.
+        input = torch.randn(5)
+        output = lowered_module.forward(input)
```
This call should fail, `forward` should take two arguments. We should have the test backend trivially implement `forward` (calling `torch::tanh` on inputs and returning it), then test that the Python, TS, and lowered module all produce the same result given an input.

SplitInfinity commented:
Made some updates, summarized below:
- switched to using `CodeTemplate` for code generation
- made `Backend::execute` take only a `Stack` and made `prim::backend_execute` varags (but not varret, more on that below) to support multiple inputs
- added plumbing to codegen methods on the lowered module that match exactly the methods on the original module specified in `method_compile_spec`
- changed underlying storage for custom class functions from `std::vector` to `std::unordered_map`, added some checking for name collisions
- got rid of `PyTorchBackendModuleInterface` due to serialization issues
- updated tests to run the module using Python, TorchScript and the test backend and compare results between all three (however, the test backend is hardcoded to do what the test `Module` does)

Now for the new problems:
- you might have noticed that I made `prim::backend_execute` vararg, but not varret. For reasons I don't yet understand, making it varret (i.e. changing the schema to `-> ...`) makes it impossible to call in Python. I left it as vararg with a single output for now but I'm not sure how to get around this.
- serialization works, but deserialization does not due to the presence of `Any` types among the members of the lowered module

SplitInfinity commented on test/jit/test_backends.py:
```diff
@@ -0,0 +1,63 @@
+
+        # Test save and load.
+        lowered_module.save("lowered.pt")
+        # loaded_module = torch.jit.load("lowered.pt")
```
So serialization works now, but deserialization doesn't work because `Any` types are not deserializable. cc: @suo

SplitInfinity commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,179 @@
+              any_dict_ty,
+              toIValue(method_compile_spec, any_dict_ty).toGenericDict());
+
+          //   This is a pointer to a backend instance that is used to access
```
TODO: Fix spaces. I thought `clang-format` would but I guess not.

SplitInfinity commented on torch/csrc/jit/runtime/register_backend_ops.cpp:
```diff
@@ -0,0 +1,77 @@
+        },
+        aliasAnalysisFromSchema()),
+    Operator(
+        "prim::backend_execute(Capsule c, Any a, ...) -> Any",
```
If you register this op as `prim::backend_execute(Capsule c, Any a, ...) -> ...`, it is not possible to call from Python. If that's what we want to do, I think some changes to how ops are located and dispatched to might be necessary.

zdevito commented on aten/src/ATen/core/custom_class.cpp:
```diff
@@ -36,13 +36,39 @@ bool isCustomClass(const c10::IValue& v) {
 
-std::vector<std::unique_ptr<jit::Function>>& customClassMethods() {
-  static std::vector<std::unique_ptr<jit::Function>> customClassMethods;
+std::unordered_map<c10::QualifiedName, std::unique_ptr<jit::Function>>& customClassMethods() {
```
I am skeptical of an approach for to_backend that requires changes to the fundamentals of custom class registration, so I am wary to allow functions like getCustomClassMethod.   This creates a ton of API confusion: now methods maybe are registered in a class, or maybe they are in this global map. Previously this list only existed to keep them alive and was not part of the public API for custom classes.

One major design point for our to_backend work is that it should just be helpful infrastructure for someone creating a backend. That is, it should guide the backend implementor to make good methods for serialization and object registration, but it should not be enabling something that wasn't already possible using custom classes.

SplitInfinity replied:
Would you be comfortable with a reintroduction of the separate registry that was present in the first version of this PR then? We cannot register these backend classes in the same way as other custom classes (and in doing so, reuse the existing class registry as-is without introducing another) because the consequent restrictions on types of custom class method arguments don't allow us to pass `IValues` back and forth between JIT and backend code (either directly or on a `Stack`) like we want to.

suo replied:
Discussed offline a bit about this: the only we need a "backend registry" (or a stealth registry like this one) is to do pybinding of the `to_backend` functions, since we need to look up all the backends to bind them one at a time.

If instead we deferred the pybinding of the that function and made the user do it (which is in-line with our desire to have the `to_backend()` functions going in the backend's py module anyway), we don't need to mess with the custom class registry afaict.

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -381,6 +381,7 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     ${TORCH_SRC_DIR}/csrc/jit/runtime/interpreter.cpp
     ${TORCH_SRC_DIR}/csrc/jit/ir/constants.cpp
     ${TORCH_SRC_DIR}/csrc/jit/ir/node_hashing.cpp
+    ${TORCH_SRC_DIR}/csrc/jit/backends/backend.cpp
```
https://github.com/pytorch/pytorch/pull/37267 :)

SplitInfinity commented on test/jit/test_backends.py:
```diff
@@ -0,0 +1,63 @@
+        self.assertEqual(jit_output, backend_output)
+
+        # Test save and load.
+        lowered_module.save("lowered.pt")
```
I got rid of this since deserialization doesn't work at the moment anyway.

SplitInfinity commented:
@jamesr66a @suo 

Added a workaround for the multiple return values problem. TL;DR - A lowered method is free to accept and return multiple values. But if it does return multiple method, the `prim::backend_execute` op packs the multiple return values into a tuple. This breaks the schema of `prim::backend_execute`, but the Python interface is exactly what you'd expect:
```
a = lowered_module.method_that_returns_one_value(...) # a is a Tensor
a, b = lowered_module.method_that_returns_two_values(...) # a and b are Tensors
```

jamesr66a commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,44 @@
+#pragma once
```
minor: I think this file should appear somewhere less deeply nested, since external users will depend on it. For example, it could be something like `torch/backend.h`, analogous to `torch/custom_class.h`

jamesr66a commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,44 @@
+namespace torch {
+namespace jit {
+// Interface for a JIT backend.
+class TORCH_API PyTorchBackendInterface : public torch::CustomClassHolder {
```
For the future: We should write a tutorial as well, I can help with that

jamesr66a commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,179 @@
+              cloned_module._ivalue(),
+              toIValue(method_compile_spec, any_dict_ty).toGenericDict());
+          preprocess_fn->run(stack);
+          auto preprocessed_module = pop(stack);
```
Do we expect the stack to be empty after this line? Should probably add an assert.

Alternatively above you could use `Function::operator()`, which takes a stack and returns the single IValue return value rather than relying on Stack pre/postconditions

jamesr66a commented on torch/csrc/jit/runtime/register_backend_ops.cpp:
```diff
@@ -0,0 +1,87 @@
+// methods for a JIT backend.
+RegisterOperators reg({
+    Operator(
+        "prim::backend_create(str a, Any a) -> Capsule",
```
Both parameters have the same name here, might mess stuff up for kwarg schema matching

I'd suggest giving these parameters meaningful names as well

jamesr66a commented on torch/csrc/jit/runtime/register_backend_ops.cpp:
```diff
@@ -0,0 +1,87 @@
+          auto method_compile_spec = pop(stack).toGenericDict();
+          auto module = pop(stack);
+          auto capsule = pop(stack).toCapsule();
+          auto backend = at::dynamic_intrusive_pointer_cast<
```
Dynamic casting might cause issues on mobile (though I remember Kimish at one point turned RTTI on for mobile). In any case, if we can make sure that we only ever get an instance of the PyTorchBackendInterface here, we might be able to use static cast. Might have to be some enforcement in our own type system, though

Probably not important to do that on this PR, just if we have issues with the dynamic cast

jamesr66a commented on torch/csrc/jit/runtime/register_backend_ops.cpp:
```diff
@@ -0,0 +1,87 @@
+
+          if (num_return_values > 1) {
+            auto return_values = pop(stack, num_return_values);
+            push(stack, c10::ivalue::Tuple::create(return_values));
```
Doesn't the schema specify the return value as a list and not tuple?

jamesr66a reviewed (commented):
Looking good! I'll probably patch your diff to try some experiments out in building some toy backends, so I'll keep everyone updated on any rough edges I run into

@suo @zdevito is this looking OK for approval?

SplitInfinity commented:
@suo and I discussed offline yesterday after I updated my PR (sorry about that!) about how to avoid cramming these backend functions into the torchbind registry and the new plan is to register backend classes using `torch::class_` but to add special `_def_unboxed` methods to get around the boxing issues we ran into a week ago so that we can pass `IValues` and stuff into them. As an added bonus, that eliminates the need for the `prim::backend_create/compile/execute` functions altogether.

Consequently, please do not review this just yet. I apologize for all this churn. :(

jamesr66a commented:
@SplitInfinity Keep in mind TorchBind classes are not serialized, so everything is going to be materializable free of context from the model

SplitInfinity commented:
Alright, just pushed another version that includes a bunch of changes discussed offline with @suo. The most notable difference is that backend classes are now registered using `torch::class_` with an extra method `_def_unboxed` added to allow the backend classes to accept args of types that normal boxed functions don't allow. This is wrapped inside a `torch::backend_` class to make registration a one-liner once the backend class has been defined. I think this is super super close to what we want, but there are still some unresolved issues:

- There is still a lookup (prefix search) into the custom class registry in `backend_init.cpp`. I think this can be fixed by adding a method on `torch::backend_` that returns the function that is currently bound in `backend_init.cpp` so that the backend writer can bind it. I will fix this in a future PR.
- There is some clowniness in `getExecuteFunction()` regarding support for multiple return values. The `preprocess`, `compile` and `execute` methods defined by the backend class are wrapped in `jit::BuiltinOpFunctions`, which can have one and only one return value. As a result, if the backend returns multiple values, they have to be packed into a tuple in order to marshal them back to Python. This is a problem caused by the reuse of `jit::BuiltinOpFunction`, but at the same time, extending `jit::Function` to remove this limitation just for this one use case seems like overkill.

suo commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,155 @@
+    // TODO: Stack does not have the number of arguments on it (probably because
+    // schema is not truly varret) so the best we can do is to assume that
+    // everything on the stack is an argument to execute.
+    auto num_args = stack.size();
```
The interpreter stack might contain other stuff (it's a stack machine so it uses the stack to hold intermediates and such), so `stack.size()` does not work; you are only supposed to pop off the arguments you know that you need. Let me look into why we're not getting the number of arguments.

jamesr66a commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include <ATen/core/stack.h>
+#include <aten/src/ATen/core/builtin_function.h>
```
remove `aten/src`

jamesr66a commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,155 @@
+      push(stack, c10::ivalue::Tuple::create(methodStack));
+    }
+  };
+}
```
Are these five functions expected to be user-visible? if not, can we put them in a [detail namespace](https://github.com/pytorch/pytorch/blob/06928047478fab33678bd9248886a938b9b07937/torch/custom_class_detail.h#L10) and potentially in a different file? I just ask because I was reading this header as if I were a new user and I was confused as to what's an external API v.s. what's an implementation

jamesr66a commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,155 @@
+
+// Static registration API for backends.
+template <class TBackendInterface>
+class backend_ {
```
Don't know if we need the underscore, `class_` uses it just because `class` is a C++ keyword :)

jamesr66a reviewed (commented):
Nice, surprised at how little actual implementation code there is now

suo commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,397 @@
+            def $method(self${,def_inputs}):
+                typed_inputs: List[Any] = [${fwd_inputs,}]
+                $ret, = self.__backend.execute(self.__handles["$method"], typed_inputs)
+                return $ret
```
you'll need to perform type refinement on the elements of `$ret`. Currently this is the generated code:
```
def forward(self,
    x: Tensor,
    h: Tensor) -> Tuple[Any, Any]:
  typed_inputs = annotate(List[Any], [x, h])
  _0 = (self.__backend).execute((self.__handles)["forward"], typed_inputs, )
  _1, _2, = _0
  return (_1, _2)
```
We have to add `isinstance(_1, Tensor)`, etc. to make sure that we are actually returning tensors.

suo replied:
[followup] the isinstance approach won't work for any container types, e.g. `List[Tensor]`, because that's not valid python. We need to figure out how to support that. Also if we're checking container values, performance becomes a concern (as we ahve to do a linear inspection of all contained elements). An alternative is to make this an implicit requirement on backends and perform an unchecked cast (i.e. we're going to assume you handed us back the right type) but we should discuss.

suo commented on torch/csrc/jit/backends/backend.h:
```diff
@@ -0,0 +1,397 @@
+      auto any_dict_ty = DictType::create(StringType::get(), AnyType::get());
+
+      // Generate LoweredModule.
+      Module loweredModule("torch.jit." + backend_name + "LoweredModule");
```
[followup] This needs to generate the module in the Python compilation unit, and properly mangle the class name. Otherwise there will be very strange behavior when you try to define more than one lowered module at a time.

suo reviewed (commented):
Very close! I made some inline comments. Ones that I marked "followup" don't have to be addressed in this PR in the interest of making sure this one doesn't spin out of proportion, but will need to be addressed after.

SplitInfinity commented:
Tracking follow-ups in #37841. I addressed the comment about refining the return values of the backend's `execute` function. I expect some Windows failures because of MSVC idiosyncrasies, so I'll wait for those CI jobs and deal with those before re-requesting review.

yinghai commented on torch/csrc/jit/backends/backend_interface.h:
```diff
@@ -0,0 +1,37 @@
+// Interface for a JIT backend.
+class TORCH_API PyTorchBackendInterface : public torch::CustomClassHolder {
+ public:
+  inline PyTorchBackendInterface() = default;
+  inline virtual ~PyTorchBackendInterface() = default;
```
```
PyTorchBackendInterface() {};
virtual ~PyTorchBackendInterface() {};
```

suo reviewed (approved):
nice! Let's land this (provided MSVC is appeased) and get started on the followups :)

AshkanAliabadi commented on torch/csrc/jit/backends/backend_interface.cpp:
```diff
@@ -0,0 +1,12 @@
+#pragma once
```
Try removing this as well.  Not sure if this might have anything to do with it.

malfet commented on torch/csrc/jit/backends/backend_interface.h:
```diff
@@ -0,0 +1,37 @@
+namespace jit {
+
+// Interface for a JIT backend.
+class TORCH_API PyTorchBackendInterface : public torch::CustomClassHolder {
```
`TORCH_API`  (which is an alias to `CAFFE2_API`) can not be used to implement the code outside of main library, because it becomes `C10_EXPORT` outside of main lib, see https://github.com/pytorch/pytorch/blob/master/c10/macros/Export.h#L96:
```
...
#define C10_HIDDEN
#define C10_EXPORT __declspec(dllexport)
#define C10_IMPORT __declspec(dllimport)
...
#ifdef CAFFE2_BUILD_MAIN_LIB
#define CAFFE2_API C10_EXPORT
#else
#define CAFFE2_API C10_IMPORT
#endif
```
I.e. when `backend_interface.cpp` is compiled as part `torch_python.dll`, `TORCH_API` becomes `__declspec(dllimport)`, which tells linker that vftable should be deinfed in `torch_cpu.dll`, which in turn leads to error you are seeing.

One way to solve the issue is to move `backend_interface.cpp` to `torch_cpu`

mortzur commented on torch/csrc/jit/backends/backend_init.cpp:
```diff
@@ -0,0 +1,11 @@
+namespace jit {
+
+void initJitBackendBindings(PyObject* module) {
+  initTestBackendBindings(module);
```
Should NewBackend include its header here and add initNewBackendBindings(..) ?

SplitInfinity replied:
No, the plan is to move `TestBackend` to the tests directory and have it register its bindings with `torch._C` or a C extension module created just for testing. 

`NewBackend` should register its bindings in its own C extension module.
